dsl/dsl.cc

```diff
@@ -112,10 +113,41 @@ class DSLReplacer {
             return classDef;
         }
 
+        auto synthesizeInitialize = false;
+        for (auto &a : classDef->ancestors) {
+            if (ChalkODMProp::isTStruct(a.get())) {
+                synthesizeInitialize = true;
+                break;
+            }
+        }
         auto oldRHS = std::move(classDef->rhs);
         classDef->rhs.clear();
-        classDef->rhs.reserve(oldRHS.size());
-
+        // the `2` is: 1 for the sig, 1 for the method
+        classDef->rhs.reserve(oldRHS.size() + (synthesizeInitialize ? 0 : 2));
+
+        if (synthesizeInitialize) {
+            // we define ours synthesized initialize first so that if the user wrote one themselves, it overrides ours.
+            auto loc = classDef->loc;
+            ast::MethodDef::ARGS_store args;
+            ast::Hash::ENTRY_store sigKeys;
+            ast::Hash::ENTRY_store sigVals;
+            args.reserve(props.size());
+            sigKeys.reserve(props.size());
+            sigVals.reserve(props.size());
+            for (auto &prop : props) {
+                auto arg = ast::MK::KeywordArg(loc, ast::MK::Local(loc, prop.name));
+                if (prop.optional) {
+                    arg = ast::MK::OptionalArg(loc, std::move(arg), ast::MK::Unsafe(loc, ast::MK::Nil(loc)));
+                }
+                args.emplace_back(std::move(arg));
+                sigKeys.emplace_back(ast::MK::Symbol(loc, prop.name));
+                sigVals.emplace_back(std::move(prop.type));
+            }
+            classDef->rhs.emplace_back(
+                ast::MK::SigVoid(loc, ast::MK::Hash(loc, std::move(sigKeys), std::move(sigVals))));
+            classDef->rhs.emplace_back(ast::MK::Method(loc, loc, core::Names::initialize(), std::move(args),
+                                                       ast::MK::EmptyTree(), ast::MethodDef::DSLSynthesized));
+        }
```

Comment:
This code should almost certainly not be in `dsl.cc`. Can we make "adding `T::Struct#initialize`" feel exactly the same as any other DSL pass?

Reply:
I see what you mean, and I was a bit uncomfortable with putting this here. But, given that `ChalkODMProp::replaceDSL` only operates on a single call to `prop`, it wasn't clear to me where this code (which I want to run "at the end") should go. Do you have thoughts on how to better structure this?

Reply:
The way I did it when I was prototyping this back in October was to use `patchDSL`:

https://git.corp.stripe.com/stripe-internal-archive/sorbet/pull/1597/files#diff-f1d169de1bf307604ed59c9b4b15859eR24

(Also instead of building up something like NodesAndProp, I just factored the prop parsing logic out of the replaceDSL method, so that I could call it from patchDSL.)

Reply:
I see. I could certainly move the generation of `initialize` to its own `patchDSL` call. However, even if I did that I'd still have to thread the information from the collected vector of `props` back into that `patchDSL` call. Should I try to do something like what you did where you parse the prop information after the fact? (The reason why I initially didn't do that here is because I feel like that would lead to duplication of logic concerning the rules of how the prop DSL works. Perhaps there's a better way to share that information than what I've come up with here?)

Reply:
Resolved in #1036.